### PR TITLE
Add/pre conditioning

### LINKED ISSUE #10: Pre-conditioning
Sort ledger by dates
Check entries in ledger balance for each transaction

### PR DESCRIPTION
**Pull request by Lebbitt**

### CODE REVIEW

File: json_journal/preconditioning.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    Ledger : dict
+        Loaded json file.
+
+    Returns
+    -------
+    Sorted_ledger : dict
+        The same json file, ordered by date.
+    '''
+    sorted_ledger = sorted(ledger, key=lambda d: d['date'])
+    return sorted_ledger
```
Comment (g-duff) [inline json_journal/preconditioning.py]: This could be a one-liner

File: json_journal/preconditioning.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    amount_sum = 0
+    for transaction in ledger:
+        entries = transaction['entries']
+
+        for entry in entries:
+            amount = entry['amount']
+            amount_sum += amount
+
+    if amount_sum == 0:
+        print("Entries sum to zero")
```
Comment (g-duff) [inline json_journal/preconditioning.py:44]: Do we consider this to be an [error](https://docs.python.org/3/tutorial/errors.html)?

File: json_journal/preconditioning.py (hunk 1 above)
Comment (Lebbitt) [reply json_journal/preconditioning.py]: done

File: json_journal/preconditioning.py (hunk 2 above)
Comment (Lebbitt) [reply json_journal/preconditioning.py:44]: done

File: json_journal/preconditioning.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    Returns
+    -------
+    Sorted_ledger : dict
+        The same json file, ordered by date.
+    '''
+    return sorted(ledger, key=lambda d: d['date'])
+
+
+def check_entries_sum(ledger):
```
Comment (g-duff) [inline json_journal/preconditioning.py]: Could this function name better represent its purpose?

File: tests/test_preconditioning.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
+# pylint: disable = import-error, missing-class-docstring, missing-function-docstring, missing-module-docstring
+import unittest
+from json_journal import preconditioning
+
+
+class TestSortLedger(unittest.TestCase):
+
+    def test_simplecase(self):
```
Comment (g-duff) [inline tests/test_preconditioning.py]: Not a very descriptive test case name

File: json_journal/preconditioning.py (hunk 3 above)
Comment (Lebbitt) [reply json_journal/preconditioning.py]: done

File: tests/test_preconditioning.py (hunk 4 above)
Comment (Lebbitt) [reply tests/test_preconditioning.py]: done

File: json_journal/preconditioning.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    Check entries in a transaction sum to zero.
+
+    Parameters
+    ----------
+    Ledger : dict
+        Dictionary with dates, descriptions and entries.
+
+    Returns
+    -------
+    Print statement
```
Comment (g-duff) [inline json_journal/preconditioning.py]: I don't think this is returned, more of a side-effect.

**Review by g-duff — CHANGES_REQUESTED** [review]

File: json_journal/preconditioning.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+'''Data sorted before being used by analysis'''
+
+
+def sort_journal_by_date(ledger):
+    '''
+    Sort JSON file by date.
```
Comment (g-duff) [inline json_journal/preconditioning.py]: Is it a JSON file?

**Review by g-duff — CHANGES_REQUESTED** [review]

File: json_journal/preconditioning.py (hunk 5 above)
Comment (Lebbitt) [reply json_journal/preconditioning.py]: removed

File: json_journal/preconditioning.py (hunk 6 above)
Comment (Lebbitt) [reply json_journal/preconditioning.py]: done

File: tests/test_preconditioning.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        ]
+
+        # When
+        actual_ledger = preconditioning.sort_journal_by_date(ledger)
+
+        # Then
+        self.assertEqual(expected_ledger, actual_ledger)
+
+
+class TestEntriesSumToZero(unittest.TestCase):
```
Comment (g-duff) [inline tests/test_preconditioning.py:75]: Is it worth adding a test for the error case?

Comment (Lebbitt) [reply tests/test_preconditioning.py:75]: done

File: tests/test_preconditioning.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+        expected_amount_sum = 0
+
+        # When
+        actual_amount_sum = preconditioning.check_entries_sum_to_zero(ledger)
+
+        # Then
+        self.assertEqual(expected_amount_sum, actual_amount_sum)
+
+class TestEntriesDoNotSumToZeroError(unittest.TestCase):
```
Comment (g-duff) [inline tests/test_preconditioning.py]: These can be the same class. I'd recommend distinguishing test cases in the test name.

`testmethod_situation()`

File: tests/test_preconditioning.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                "date": "2022-11-10",
+                "description": "savings",
+                "entries": [
+                    {"account": "assets:savings", "amount": 50},
+                    {"account": "assets:bank", "amount": -50}
+                ]
+            }
+        ]
+
+        # Then
```
Comment (g-duff) [inline tests/test_preconditioning.py]:
```suggestion
        # When/Then
```

File: tests/test_preconditioning.py (hunk 8 above)
Comment (Lebbitt) [reply tests/test_preconditioning.py]: done

File: tests/test_preconditioning.py (hunk 9 above)
Comment (Lebbitt) [reply tests/test_preconditioning.py]: done

**Review by g-duff — APPROVED** [review]

**Pull request merged**